asserts/tool/sign.go

```
@@ -0,0 +1,221 @@
+	AuthorityID string
+
+	// the assertion type (as a string)
+	AssertionType string
+	// specify the media type of the input
+	StatementMediaType string
+	// Statement is used as input to construct the assertion
+	// it's a mapping encoded as either JSON or YAML (specified in StatementMediaType)
+	// either of just the header fields of the assertion
+	// or containting exactly two entries
+	// "headers": mapping with the header fields
+	// "content-body": used as the content body of the assertion
+	Statement []byte
+
+	// revision of the new assertion
+	Revision int
+}
+
+func parseStatement(req *SignRequest, dest interface{}) error {
+	switch req.StatementMediaType {
+	case YAMLInput:
+		err := yaml.Unmarshal(req.Statement, dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as YAML: %v", err)
+		}
+	case JSONInput:
+		dec := json.NewDecoder(bytes.NewBuffer(req.Statement))
+		// we want control over supporting only integers
+		dec.UseNumber()
+		err := dec.Decode(dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as JSON: %v", err)
+		}
+	default:
+		return fmt.Errorf("unsupported media type for assertion input: %q", req.StatementMediaType)
+	}
+	return nil
+}
+
+type nestedStatement struct {
+	Headers     map[string]interface{} `yaml:"headers" json:"headers"`
+	ContentBody string                 `yaml:"content-body" json:"content-body"`
+}
+
+// Sign produces the text of a signed assertion as specified by req.
+func Sign(req *SignRequest, keypairMgr asserts.KeypairManager) ([]byte, error) {
+	typ := asserts.Type(req.AssertionType)
+	if typ == nil {
+		return nil, fmt.Errorf("invalid assertion type: %q", req.AssertionType)
+	}
+	if req.Revision < 0 {
+		return nil, fmt.Errorf("assertion revision cannot be negative")
+	}
+	if req.AccountKey == nil && req.KeyID == "" {
+		return nil, fmt.Errorf("both account-key and key id were not specified")
+	}
```

It should return an error if AccountKey is provided *and* either KeyID or AccountID are non-zero. But I sort-of feel this struct is a bit overloaded right now ... especially as the AccountID can also be in the headers.